File: playgrounds/bytecode/bytecode.js
function setHealth(wizardId, amount) {}
function setWisdom(wizardId, amount) {}
function setAgility(wizardId, amount) {}
function playSound(soundId) {}
function spawnParticles(particleType) {}

const instructionEnum = {
  setHealth: "setHealth",
  setWisdom: "setWisdom",
};

class VM {
  constructor() {
    this.stack = [];
    this.size = 0;
  }

  push(value) {
    this.stack.push(value);
    this.size += 1;
  }

  pop() {
    if (this.size > 0) {
      return this.stack[size];
    } else {
      throw new Error("stack does not have value");
    }
  }
  // stack : 
  interpret(bytecode = []) {
    for(let index = 0; index < this.size) {
      const instruction = bytecode[index];
      switch (instruction) {
        case instructionEnum.setHealth: {
          const amount = this.pop(); // 10
          const wizard = this.pop(); // 0
          setHealth(wizard, amount);
          break;
        }
        case instructionEnum.setWisdom: {
          const wizard = this.pop();
          const amount = this.pop();
          setWisdom(wizard, amount);
          break;
        }
        case INSTRUCTION_LITERAL: {
          index += 1;
          this.push(bytecode[index]);
          break;
        }
      }
    }
  }
}

File: playgrounds/bytecode/bytecode.rs
fn setHealth(wizardId: usize, amount: i32);
fn setWisdom(wizardId: usize, amount: i32);
fn setAgility(wizardId: usize, amount: i32);
fn playSound(soundId: usize);
fn spawnParticles(particleType: ParticleType);


enum ParticleType {}

enum Instruction {
    SetHealth(i32, i32),
    SetWisdom(i32, i32),
}

enum CustomError {
    StackOverflow,
    StackUnderflow
}

struct VM {}

impl VM {
    pub fn interpret(&mut self, bytecodes: [Instruction]) -> Result<(), CustomError> {
        for bytecode in bytecodes {
            match bytecode {
                Instruction::SetHealth(wizardId, value) => setHealth(wizardId, value),
                Instruction::SetWisdom(wizardId, value) => setWisdom(wizardId, value),
            };
        }

        Ok(())
    }
}